### PR TITLE
issues of VPNService.kt

### PR DESCRIPTION
**Pull request by rurirei**
https://github.com/2dust/v2rayNG/issues/609

### CODE REVIEW

File: V2rayNG/app/src/main/kotlin/com/v2ray/ang/service/V2RayVpnService.kt [hunk 1]
```diff
@@ -200,9 +200,12 @@ class V2RayVpnService : VpnService(), ServiceControl {
 //        val emptyInfo = VpnNetworkInfo()
 //        val info = loadVpnNetworkInfo(configName, emptyInfo)!! + (lastNetworkInfo ?: emptyInfo)
 //        saveVpnNetworkInfo(configName, info)
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P && listeningForDefaultNetwork) {
-            connectivity.unregisterNetworkCallback(defaultNetworkCallback)
-            listeningForDefaultNetwork = false
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+            try {
+                connectivity.unregisterNetworkCallback(defaultNetworkCallback)
```
Comment (yuhan6665) [inline V2rayNG/app/src/main/kotlin/com/v2ray/ang/service/V2RayVpnService.kt:211]: I feel like we can simply check null for defaultNetworkCallback then we don't need try catch?

File: V2rayNG/app/src/main/kotlin/com/v2ray/ang/service/V2RayVpnService.kt [hunk 2]
```diff
@@ -153,8 +151,11 @@ class V2RayVpnService : VpnService(), ServiceControl {
         }
 
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
-            connectivity.requestNetwork(defaultNetworkRequest, defaultNetworkCallback)
-            listeningForDefaultNetwork = true
+            try {
+                connectivity.requestNetwork(defaultNetworkRequest, defaultNetworkCallback)
+            } catch (ignored: Exception) {
+                // ignored
```
Comment (yuhan6665) [inline V2rayNG/app/src/main/kotlin/com/v2ray/ang/service/V2RayVpnService.kt:157]: Let's not ignore this exception by logging it. By the way, if you know the exact exception, we should see if there are other ways to fix the root problem.

**Review by yuhan6665 — COMMENTED** [review]
Thanks for your pull request! left two small comments.

File: V2rayNG/app/src/main/kotlin/com/v2ray/ang/service/V2RayVpnService.kt (hunk 1 above)
Comment (rurirei) [reply V2rayNG/app/src/main/kotlin/com/v2ray/ang/service/V2RayVpnService.kt:211]: catch is needed for avoiding crash. the reason of IOException is something i do not care (actually Android has too much addition like this). exception at registerDefaultNetwork could be simply ignored even now.

Comment (yuhan6665) [reply V2rayNG/app/src/main/kotlin/com/v2ray/ang/service/V2RayVpnService.kt:211]: Fine, we don't care for unregister(). But we should care for register(). We have this code for a reason and some users are depending on it. We should at least add a log for exception. See more details
https://github.com/shadowsocks/shadowsocks-android/commit/3a8f863bd1f0deed081de082a674020368f30319

Comment (rurirei) [reply V2rayNG/app/src/main/kotlin/com/v2ray/ang/service/V2RayVpnService.kt:211]: mostly `try` is needed, catch IOException and analysis is maybe a way.

**Review by rurirei — COMMENTED** [review]
third commit:

call sendFd() before tun2socks started leads to java.io.IOException: connection refused at localSocket() ..
https://github.com/2dust/v2rayNG/blob/df3f1ca3efdd5b6e9115cbe7ecdb4260541ab732/V2rayNG/app/src/main/kotlin/com/v2ray/ang/service/V2RayVpnService.kt#L178
even sendFd() called twice at https://github.com/2dust/v2rayNG/blob/df3f1ca3efdd5b6e9115cbe7ecdb4260541ab732/AndroidLibV2rayLite/interact.go#L240 https://github.com/2dust/v2rayNG/blob/df3f1ca3efdd5b6e9115cbe7ecdb4260541ab732/V2rayNG/app/src/main/kotlin/com/v2ray/ang/service/V2RayVpnService.kt#L166
so it needs called only once after escort.run() hangs tun2socks up.

**Pull request merged**